File: lib/prim.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = long long;

struct Edge {
  ll to;
  ll cost;
};
using Graph = vector<vector<Edge>>;

// 最小全域木を求めるアルゴリズム
// 計算量: O(ElogV)
// 既に到達した頂点集合からまだ到達していない頂点集合への辺のうち、コストが最小のものを選んでいく
Graph prim(Graph &G) {
  Graph ret(G.size());
  vector<bool> used(G.size(), false);
  auto c = [](Edge l, Edge r) { return l.cost > r.cost; };
  // コストが小さいものから取り出す
  // decltype(式): 式の型を返す
  priority_queue<Edge, vector<Edge>, decltype(c)> que(c);
  que.push(Edge{0, 0});
  while (!que.empty()) {
    auto e = que.top();
    que.pop();
    ll v = e.to;
    // 一回行ったところはスキップ
    if (used[v]) {
      continue;
    }
    used[v] = true;
    ret[0].push_back(e);
    // 次のノードをすべてキューにいれる
    for (ll i = 0; i < G[v].size(); i++) {
      que.push(G[v][i]);
    }
  }

  return ret;
}

// http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=GRL_2_A&lang=jp
int main() {
  int V, E;
  cin >> V >> E;

  Graph G(V);

  for (int i = 0; i < E; i++) {
    int si, ti, wi;
    cin >> si >> ti >> wi;
    G[si].push_back(Edge{ti, wi});
    G[ti].push_back(Edge{si, wi});
  }

  auto t = prim(G);

  // コストの総和
  ll ans = 0;
  for (ll i = 0; i < t.size(); i++) {
    for (ll j = 0; j < t[i].size(); j++) {
      ans += t[i][j].cost;
    }
  }
  cout << ans << endl;
  return 0;
}

File: lib/prim.py
import heapq


# 隣接リストによるグラフ表現
Graph = list[list[tuple[int, int]]]


def prim(G: Graph) -> Graph:
    """最小全域木を求めるアルゴリズム

    Args:
        G (Graph): グラフ

    Returns:
        Graph: 最小全域木
    """
    ret = [[] for _ in range(len(G))]
    used = [False] * len(G)
    que = []
    heapq.heappush(que, (0, 0))
    while que:
        e = heapq.heappop(que)
        v = e[1]
        # 一回行ったところはスキップ
        if used[v]:
            continue
        used[v] = True
        ret[0].append((e[1], e[0]))
        # 次のノードをすべてキューにいれる
        for i in range(len(G[v])):
            que.append((G[v][i][1], G[v][i][0]))
            heapq.heapify(que)

    return ret


# http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=GRL_2_A&lang=jp
if __name__ == "__main__":
    V, E = map(int, input().split())
    G = [[] for _ in range(V)]

    for i in range(E):
        si, ti, wi = map(int, input().split())
        G[si].append((ti, wi))
        G[ti].append((si, wi))

    t = prim(G)

    # コストの総和
    ans = 0
    for i in range(len(t)):
        for j in range(len(t[i])):
            ans += t[i][j][1]
    print(ans)
